support condition objects in utility functions #6666

jharting opened the pull request:
### Pull Request check-list

_Please make sure to review and check all of these items:_
- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
  - I did not with mssql (don't have it available)
  - linting seems broken in master (13k problems)
- [x] Does your issue contain a link to existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Have you added new tests to prevent regressions?
- [x] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
  - not yet, want to hear team's opinion on the change first
- [x] Have you added an entry under `Future` in the changelog?

_NOTE: these things are not required to open a PR and can be done afterwards / while the PR is open._
### Description of change

https://github.com/sequelize/sequelize/issues/6666

Makes it possible to pass condition objects (e.g. `{foo: 'bar'}`) to `Sequelize.fn()`, `Sequelize.cast()`.
This makes it easy to use conditional counts in Sequelize queries, e.g.

`Sequelize.fn('sum', {name: 'Fred'})` which results in `SUM(``name``= 'Fred')` which can be used in attribute definition to count entities for which the condition matches.

I did not get a chance to test with mssql as I do not have it available. If anyone can do that for me I would appreciate that.

janmeier commented on test/unit/utils.test.js:
```diff
@@ -456,4 +457,25 @@ suite(Support.getTestDialectTeaser('Utils'), function() {
+    var run = generator.handleSequelizeMethod.bind(generator);
+    var expectsql = Support.expectsql
+
+    if (Support.getTestDialect() !== 'mssql') {
```
No need to skip mssql - It doesn't need to be 100 percent correct syntax (in my mind), as long as it tests the functionality (that you can pass nested conditions)

janmeier commented on test/unit/utils.test.js:
```diff
@@ -456,4 +457,25 @@ suite(Support.getTestDialectTeaser('Utils'), function() {
+            bar: 'bar'
+          }
+        }, 'int'))), {
+          default: 'SUM(CAST((`foo` = \'foo\' OR `bar` = \'bar\') AS INT))',
```
Default should use `[` and `]` for escaping

janmeier commented on test/integration/utils.test.js:
```diff
@@ -226,4 +228,80 @@ describe(Support.getTestDialectTeaser('Utils'), function() {
     });
   });
+
+  describe('Sequelize.fn', function() {
```
While an integration test might not be strictly needed, I think this is good to keep as an example of how the functionality can be used

felixfbecker commented:
You can test MSSQL through AppVeyor ;)

jharting commented:
@janmeier @felixfbecker Thanks for feedback! I addressed some of the comments. I still do not have access to mssql (not even sure if it supports the syntax I am using in the tests). Let me try if I can use appveyor CI runner here to get the tests pass easily but maybe not.

jharting commented:
I played with mssql and it does not seem to support this syntax. The closest alternative is `SUM(CASE WHEN foo THEN 1 ELSE 0)`. Therefore, I put mssql exclusions into the integration tests.

felixfbecker commented:
Don't forget to mention in the documentation then that this is not supported by MSSQL

jharting commented:
I modified changelog and added mention to docs. From my POV this change is now finished. Let me know if you find anything missing.

janmeier commented:
Thanks :+1:

jharting commented:
@janmeier Thank you! Btw is there a chance to get this into v3 as well? Should I open a new MR against v3? Do you plan further 3.x releases?

janmeier commented:
Definitely - Let me see if I can cherry-pick it really quick

janmeier commented:
https://github.com/sequelize/sequelize/commit/75bd5256ad53ba98d1dd535c2f2a0feaa8412d5a, just going to wait for travis to do his thing before i push a new version

janmeier commented:
`v3.27.0
